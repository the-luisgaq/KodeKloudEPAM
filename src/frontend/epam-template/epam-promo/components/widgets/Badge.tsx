import { createSkinComponent } from '@epam/uui-core';
import * as uui from '@epam/uui';
import { EpamPrimaryColor } from '../types';

type BadgeColor = EpamPrimaryColor | 'yellow'| 'orange' | 'fuchsia' | 'purple' | 'violet' | 'cobalt' | 'cyan' | 'mint'
| 'white' | 'gray10'| 'gray30'| 'gray60' | uui.BadgeProps['color'];

type BadgeMods = {
    /**
     * Defines component color.
     * @default 'blue'
     */
    color?: BadgeColor
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: uui.BadgeProps['fill'];
    /**
     * Defines component size.
     * @default 36
     */
    size?: uui.BadgeProps['size'];
};

/** Represents the properties of a badge component. */
export type BadgeProps = uui.BadgeCoreProps & BadgeMods;

export const Badge = createSkinComponent<uui.BadgeProps, BadgeProps>(
    uui.Badge,
    (props) => {
        return {
            color: props.color || 'blue',
            fill: props.fill || 'solid',
        };
    },
);
