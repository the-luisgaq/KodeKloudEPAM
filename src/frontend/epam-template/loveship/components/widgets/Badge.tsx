import { createSkinComponent } from '@epam/uui-core';
import * as uui from '@epam/uui';
import { EpamAdditionalColor, EpamPrimaryColor } from '../types';

const DEFAULT_SIZE = '18';

type BadgeColor = EpamPrimaryColor | EpamAdditionalColor | 'yellow'| 'orange' | 'purple' | 'cyan' | 'mint' | 'white'
| 'night100' | 'night300' | 'night600' | uui.BadgeProps['color'];

interface BadgeMods {
    /**
     * Defines component color.
     * @default 'sky'
     */
    color?: BadgeColor
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: uui.BadgeProps['fill'];
    /**
     * Defines component size.
     * @default '18'
     */
    size?: uui.BadgeProps['size'];
}

/** Represents the properties of a Badge component. */
export type BadgeProps = uui.BadgeCoreProps & BadgeMods;

export const Badge = createSkinComponent<uui.BadgeProps, BadgeProps>(
    uui.Badge,
    (props) => {
        return {
            color: props.color || 'sky',
            size: props.size || DEFAULT_SIZE,
            fill: props.fill || 'solid',
        };
    },
);
