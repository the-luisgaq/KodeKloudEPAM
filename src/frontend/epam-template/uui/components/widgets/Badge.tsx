import React from 'react';
import {
    Icon, IDropdownToggler, IHasCaption, IHasIcon, Overwrite, uuiElement,
} from '@epam/uui-core';
import { Clickable, ClickableComponentProps, IconContainer } from '@epam/uui-components';
import { CountIndicator } from './CountIndicator';

import { settings } from '../../settings';
import css from './Badge.module.scss';

const DEFAULT_FILL = 'solid';

type BadgeMods = {
    /** Defines component color. */
    color?: 'info' | 'success' | 'warning' | 'critical' | 'neutral';
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: 'solid' | 'outline';
    /**
     * Defines component size.
     * @default '36'
     */
    size?: '18' | '24' | '30' | '36' | '42' | '48';
};

export interface BadgeModsOverride {}

export interface BadgeCoreProps extends ClickableComponentProps, IDropdownToggler, IHasIcon, IHasCaption {
    /** Pass true to display an indicator. It shows only if fill = 'outline'. */
    indicator?: boolean;
    /**
     * Position of the icon (left of right)
     * @default 'left'
     */
    iconPosition?: 'left' | 'right';
    /** Icon for drop-down toggler */
    dropdownIcon?: Icon;
    /** Count value to be placed in component */
    count?: React.ReactNode;
}

/** Represents the properties of a Badge component. */
export interface BadgeProps extends BadgeCoreProps, Overwrite<BadgeMods, BadgeModsOverride> {}

function applyBadgeMods(mods: BadgeProps) {
    return [
        'uui-badge',
        css.root,
        `uui-size-${mods.size || settings.badge.sizes.default}`,
        `uui-fill-${mods.fill || DEFAULT_FILL}`,
        `uui-color-${mods.color || 'info'}`,
    ];
}

export const Badge = React.forwardRef<HTMLButtonElement | HTMLAnchorElement | HTMLSpanElement, BadgeProps>((props, ref) => {
    const styles = [applyBadgeMods(props), props.cx];

    const DropdownIcon = props.dropdownIcon ? props.dropdownIcon : settings.badge.icons.dropdownIcon;

    return (
        <Clickable
            { ...props }
            rawProps={ {
                'aria-haspopup': props.isDropdown,
                'aria-expanded': props.isOpen,
                ...props.rawProps,
            } }
            cx={ styles }
            ref={ ref }
        >
            {(props.indicator && props.fill === 'outline') && (
                <div className="uui-indicator"></div>
            )}
            { props.icon && props.iconPosition !== 'right' && (
                <IconContainer
                    icon={ props.icon }
                    onClick={ !props.isDisabled ? props.onIconClick : undefined }
                />
            ) }
            { props.caption && (
                <div className={ uuiElement.caption }>
                    { props.caption }
                </div>
            ) }
            { props.count !== undefined && props.count !== null && (
                <CountIndicator
                    key="count-indicator"
                    color={ null }
                    size={ settings.badge.sizes.countIndicatorMap[props.size || settings.badge.sizes.default] }
                    caption={ props.count }
                />
            ) }
            { props.icon && props.iconPosition === 'right' && (
                <IconContainer icon={ props.icon } onClick={ !props.isDisabled ? props.onIconClick : undefined } />
            ) }
            { props.isDropdown && (
                <IconContainer icon={ DropdownIcon } flipY={ props.isOpen } />
            )}
        </Clickable>
    );
});
