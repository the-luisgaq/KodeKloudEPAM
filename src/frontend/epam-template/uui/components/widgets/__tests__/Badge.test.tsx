import React from 'react';
import { Badge } from '../Badge';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';
import { ReactComponent as CalendarIcon } from '@epam/assets/icons/action-calendar-fill.svg';

describe('Badge', () => {
    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(<Badge />);
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with props', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <Badge
                caption="Test badge"
                fill="solid"
                color="success"
                icon={ CalendarIcon }
                count={ 12 }
                onIconClick={ () => {} }
                onClick={ () => {} }
                isDropdown={ true }
                size="36"
            />,
        );
        expect(tree).toMatchSnapshot();
    });
});
