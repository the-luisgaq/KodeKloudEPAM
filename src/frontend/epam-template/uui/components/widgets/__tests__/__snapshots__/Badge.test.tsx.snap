// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should be rendered correctly 1`] = `
<DocumentFragment>
  <span
    class="uui-button-box uui-enabled uui-badge root uui-size-36 uui-fill-solid uui-color-info"
    tabindex="-1"
  />
</DocumentFragment>
`;

exports[`Badge should be rendered correctly with props 1`] = `
<DocumentFragment>
  <button
    aria-haspopup="true"
    class="uui-button-box uui-enabled -clickable uui-badge root uui-size-36 uui-fill-solid uui-color-success"
    tabindex="0"
    type="button"
  >
    <div
      class="container uui-icon uui-enabled -clickable"
    >
      <svg
        class=""
      />
    </div>
    <div
      class="uui-caption"
    >
      Test badge
    </div>
    <div
      class="root uui-count_indicator uui-size-18"
    >
      12
    </div>
    <div
      class="container uui-icon uui-enabled"
    >
      <svg
        class=""
      />
    </div>
  </button>
</DocumentFragment>
`;
